components/camel-aws/src/main/java/org/apache/camel/component/aws/s3/client/AWSS3ClientFactory.java

```diff
@@ -0,0 +1,37 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.component.aws.s3.client;
+
+import org.apache.camel.component.aws.s3.S3Configuration;
+import org.apache.camel.component.aws.s3.client.impl.StandardAWSS3ClientImpl;
+import org.apache.camel.component.aws.s3.client.impl.IAMOptimizedAWSS3ClientImpl;
+
+/**
+ * Factory class to return the correct type of AWS S3 aws.
+ */
+public final class AWSS3ClientFactory {
+
+	/**
+	 * Return the correct aws s3 client (based on remote vs local).
+	 * @param maxConnections max connections
+	 * @return AWSS3Client
+	 */
+	public static AWSS3Client getAWSS3Client(S3Configuration configuration, int maxConnections) {
```

Comment:
Maybe use a client Factory only for adding a call to a builder a method is a bit overkilling. The code in two classes is almost the same except for minor things. Maybe we can refactor next.

Reply:
Hmm... Im very conflicted on this one. This decision will likely be a bit opinionated because there are pros and cons to each approach. I like the factory because its incredibly clean and easier to debug, but I also like the idea that you could have this all in one class (since it will likely only have 1 method (i.e. get client) and the differences aren't extreme by any means (not to say that there aren't significant differences)).

To me, this is more a shoulder shrug than anything. I prefer the factory because I am used to writing classes with this pattern in situations similar to this. I do see your point in regards to the refactoring/adding features to the client that may cause confusion when trying to do in multiple classes.